### PR TITLE
Concentric circle

### PR DESCRIPTION
**Pull request by lee7632**
The concentric circle mesh C file is completed. It is for generating mesh for typical LWR fuels. 
- 'Windmill' approach might need to be applied to produce refine meshes at the corners of the unit cell. 
@snschune @YaqiWang @friedmud

### CODE REVIEW

**Review by friedmud — CHANGES_REQUESTED** [review]
This is a really great start!  But there is definitely more to do before we merge it.

1.  It needs to be generalized.  I understand why you started with fuel/gap/clad... but for this to go into MOOSE it needs to just be a list of radii (an arbitrary number of them)
2.  We need to be able to generate any portion of the unit square.  We started with the quarter cell because it is easiest to think about - but now that you have that going... let's make a `MooseEnum` called `portion` that takes as options: `full top_right top_left bottom_left bottom_right right_half top_half left_half bottom_half`... with the default being `full`.  I've thought a lot about whether we wanted to just generate the quarter and flip/rotate/stitch it - or just generate the piece that we want... and I think that we want to just generate the piece that we want directly.
3.  Volume preserving should be optional (just add a `bool` parameter to turn it on and off)
4.  I can make it segfault with some combinations of options.  For instance: `num_sectors=6, nr=4, num_intervals_unit_cell=3`.  Are these cases that need error messages - or is there a logic problem?

Again: REALLY great start... but let's keep going!

**Comment by friedmud** [comment]
Oh: one more... it would be great if we could generate this _without_ the moderator... i.e. just the circles.

**Comment by snschune** [comment]
For completeness for the TODO list:
the current strategy creates large elements in the corners of the square. For LWR analysis these cause a bad representation of the scattering source leading to generally inaccurate answers. French LWR analysis uses a [windmill approach](https://www.researchgate.net/publication/270898751_New_Modelling_of_LWR_Assemblies_using_the_APOLLO2_Code_Package?_sg=GXO4Tyi5AdtuTBR5S5U7n5BMHBFM1V1dure3xv1y-tRwoeuEK-2h10KfrDJlYXAivKBXivKncg). This is fairly LWR specific and might need to go into a derived class in RSN. Even for general purpose the corner elements should be meshed a bit finer.

**Comment by friedmud** [comment]
@snschune ummm - those "meshes" don't look like meshes.

The current method has a variable size for the elements in the "square" in the middle - so you can make it as fine as you want it.

**Comment by snschune** [comment]
But it would invariable lead to larger elements in the corner away from the center (if you do the upper right quarter, then the elements adjacent go the upper right corner are large). The windmill approach simply cuts of a triangle adjacent to the corners by a diagonal line - it's the French sanctioned pin cell meshing scheme as far as I know. The depicted meshes may not be a great example though.

**Comment by brianmoose** [comment]
@lee7632 An invitation has been sent to you to join the "MOOSE Users" github team. Once you accept the invitation then civet testing will happen automatically without a developer activating them manually.

**Comment by friedmud** [comment]
@snschune I don't understand what you're saying.  Let's get together and chat about it tomorrow.

**Comment by lee7632** [comment]
@friedmud @snschune @YaqiWang 

I have addressed all four comments that Derek made. The windmill approach is currently not added, but adding the moderator part is optional. Users must provide the list of radii. In this way, mesh refinement on particular regions can be easily performed. MooseEnum called portion is added. BoundaryIDs and their names have been fixed. In addition, the volume preserving function is optional. Its documentation will be completed sometime today.

**Comment by YaqiWang** [comment]
@lee7632 can you attach few sample pictures of this mesh with different numbers of radii and different radii?

**Comment by brianmoose** [comment]
@lee7632  I resent the invitation, you should get an email. If not, you can try https://github.com/idaholab/moose/invitations or it might show up on https://github.com/idaholab/moose

**Comment by lee7632** [comment]
@brianmoose @permcody 
I have received your email and accepted the invitation. Thank you!

**Comment by lee7632** [comment]
@YaqiWang Here are some sample pictures. 

screen shot 2018-07-13 at 9 12 41 am
screen shot 2018-07-13 at 9 19 21 am
screen shot 2018-07-13 at 9 21 05 am
screen shot 2018-07-13 at 9 24 26 am
screen shot 2018-07-13 at 10 29 45 am
screen shot 2018-07-13 at 10 30 26 am
screen shot 2018-07-13 at 10 30 52 am
screen shot 2018-07-13 at 10 31 19 am
screen shot 2018-07-13 at 10 32 00 am
screen shot 2018-07-13 at 10 32 22 am
screen shot 2018-07-13 at 10 35 56 am

**Comment by YaqiWang** [comment]
Thanks @lee7632 . I see few problems but I also do not have good suggestions on solution:
1. the elements at outer rings have larger area than the inner elements;
2. elements outsize the rings at the square corner have larger areas than others;
3. aspect ratio of elements in thin rings is not good.

@friedmud showed me few nice pin mesh generated by Cubit. Can we somehow borrow the ideas from there to add the nodes? I know it is hard because the arbitrary distribution of radii.

**Comment by friedmud** [comment]
These look awesome!  Great job!

**Comment by YaqiWang** [comment]
Right, we can distribute rings with equal volumes for 1. For 2, we can add more nodes close to the four corners. Aspect ratio is defined as the max ratio of side areas and element volume. If we have thin rings, I guess we have to live with elements with large aspect ratios.

**Comment by friedmud** [comment]
This code is just want it needs to be.  The connectivity is perfect.  We
can use smoothing algorithms (either explicit - or run something like a
Laplacian smoother) to get better gradation within and without the circles
(which is how Cubit gets to its meshes)... but that's beyond the scope of
this work.

Derek

**Comment by YaqiWang** [comment]
It will be nice to have that smoother. Yes setting up connectivity is more important. Can we adjust the connectivity at the four corners so that if the smoother is in place, we will have a better mesh?

**Comment by friedmud** [comment]
I don't know which corners you're worried about.  Check out the code and play with the options first... even as it is it can generate some nice meshes.

The biggest trouble area is the transition from the square to the circles.  That can be easily smoothed out with a little more work.

**Comment by YaqiWang** [comment]
I am referring the corners related with the windmill approach @snschune mentioned.

**Comment by friedmud** [comment]
The "windmill" is not a good idea - it doesn't really solve the large element in the corner problem... and it generates a triangle.

The "windmill" paper is not generating  a mesh - it's solid geometry.

That said: I just tried out the code and we are missing some options @lee7632 .

Look at this mesh:

screen shot 2018-07-13 at 2 38 38 pm

Here are some issues:
1.  There is no way to set the size of the inner "square"
2.  There is no way to set the number of "rings" in the inner circle 
3.  We should be able to set the number of rings (radial subdivisions) in each circle - either with equal spacing or equal volume
4.  We should be able to set the number of rings in the outer part (again with equal spacing or equal volume).

Right now you can achieve (3) and (4) manually... but it's a pretty big pain (and you have to do the calculations yourself.  (2) would be possible right now if (1) were implemented... but there should be some shortcut syntax for it anyway.

**Comment by friedmud** [comment]
BTW: I plan on working on the "smoothing" issue... I talked to @jwpeterson about it this afternoon and it might be possible to do with libMesh's built in smoothers without too much trouble.

**Comment by YaqiWang** [comment]
How about we do this?
a
Basically add a small square at the corners.

**Comment by lee7632** [comment]
@friedmud @YaqiWang @snschune 
Can we have a meeting sometime next week to discuss about the details of the code built and additional modification that I need to make? 
It would be great to have a good plan that everyone can agree on and pursue through.

**Comment by friedmud** [comment]
Here's what the laplacian smoother does:

screen shot 2018-07-13 at 4 00 56 pm

It look ok - now I just need to make it honor block boundaries...

**Comment by YaqiWang** [comment]
Some of nodes (hard nodes for preserving volumes) should be never moved, some can. The nodes for the grid inside the innermost ring are the example of soft nodes, but those for drawing the rings are hard. If we have subdivisions on rings, those nodes on the subdivisions are also soft. I still believe we should add few more soft nodes close to the corners to give more freedom to the smoother. The smoothed mesh does look a lot better.

**Comment by friedmud** [comment]
Right - which why I said: "now I just need to make it honor block boundaries..."

I already did it too.  Here is what the mesh looks like now:

screen shot 2018-07-13 at 4 58 39 pm

This will be going in as a new mesh modifier (but we can also just call the function from `ConcentricCircleMesh` too)

**Comment by YaqiWang** [comment]
This looks really nice! We will see how our discretization error will be and AMG solver will perform on these meshes. If the smoother can do smoothing with volume constraints, we can make all nodes soft. We can create second order mesh easily.

**Comment by friedmud** [comment]
I agree - Let’s have a meeting to discuss all of this early next week.

**Comment by lee7632** [comment]
@friedmud 

Derek, would you mind helping me to solve the problem with 'test debug' ? I am not sure what errors mean. I tried fixing it, but not much progress.

Thank you, 
Jieun

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: use `std::vector<T>::back()`

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: use `unsigned int`, `index >= 0`, right?

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: remove brackets in else if branch.

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: remove all brackets here.

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: good start here with Quad4, but we should also think how to accommodate second order meshes. Any thoughts on that?

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: remove this blank line.

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: remove `{`

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: you already have a reference to the `MeshBase` object called `mesh`. Couldn't you simply have `mesh` be of type `ReplicatedMesh` and cast it first thing in the `build` function. That makes this cast here redundant.

File: framework/src/mesh/ConcentricCircleMesh.C
Comment (snschune) [inline framework/src/mesh/ConcentricCircleMesh.C]: also why do you need to call this when you have already called it on the cast `ReplicatedMesh` objects which are the exact same objects?

**Review by snschune — CHANGES_REQUESTED** [review]

**Comment by friedmud** [comment]
Is everyone ok with where this is now?

I think it's good enough to go in like it is.  Anything more that we want to do to it we can do later.

@lee7632 Are you happy with the way it is right now - or do you have something more that you're working on?

**Comment by snschune** [comment]
👍

**Comment by lee7632** [comment]
@ friedmud @snschune @YaqiWang I think its good to go for now! 

I am currently looking into the smoothing technique, refinement on moderator, and some test cases with rattlesnake. I talked to Mark last week, and I might be able to work on it more when I go to Texas A&M. I believe if there are more updates in future, we can incorporate them later. 

Thank you, everyone!

**Comment by YaqiWang** [comment]
This is very good to me. We definitely can grow this mesh generation capability in MOOSE and applications later. One thing I was thinking is to allow users assign data to elements easily, for example, depletion zone ID. I think if each of this mesher can have an interface like `std::vector<const Elem *> parts(const MultiMooseEnum & allowed_region_identifiers)`. Anyway, this is a really good start point.

**Comment by friedmud** [comment]
Ok - I've looked through the code and I'll merge it once mooseframework.org comes back online.

**Review by friedmud — APPROVED** [review]

**Pull request merged**
